Feature/landslide

Evelyn-M opened the pull request:
Hello both.

Here's my new pull request.

It's still about the landslide feature, no surprises :D. I removed the util function again, turned out after thorough re-consideration that there was no need for it.

So please review my module `hazard/landslide.py`, with the 3 functions it provides, as well as the corresponding tests and the tutorial.

A lot of the file changes are from deleting large test files, removing hard-coded file paths (both in `config` and `__init__()`)from previously uploaded files (which are now up to the user to download themselves).

Thanks for reviewing ;) !!

ThomasRoosli commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +46,174 @@ def __init__(self):
-                         'generation as ls_model=str')
-            raise KeyError
+        for i, j in zip(*self.fraction.nonzero()):
+            self.fraction[i, j] = binom.rvs(n=n_years,
```
The mathematics of fraction and/or frequency calculation should be checked. I think it might be `binom.rvs(n=n_years,p=self.fraction[i, j])/n_years`

Evelyn-M replied:
As is in the current code, they are still integer values (number of "successes" or landslide occurrences per grid cell over the whole time period). As the whole grid cell is either unaffected (0) or one to several times, this seemed to be rasonable in the fraction.

In `self.frequency`, I then divide the fraction value by `n_years`. 

So this should be consistent for the impact calculation, right?

ThomasRoosli commented on climada/util/coordinates.py:
```diff
@@ -34,6 +34,7 @@
 from iso3166 import countries as iso_cntry
 import numpy as np
 import pandas as pd
+import pyproj
```
delete this line

ThomasRoosli reviewed (commented):
Great work to integrate two different landslide datasets into CLIMADA that are very different from the event based mental model of a hazard, that CLIMADA generally uses.
As the two resulting hazards have a very different structure I would either split the current code into two classes: e.g. LandslidePoint, that reads in the historic dataset and LandslideGrid, that reads in the probabilistic dataset.
on the other hand: it is maybe worthwhile to invest a little more in a restructuring of the class to combine the strength of the existing code and produce a class where the historic and probabilistic Landslide hazard have a similar structure:
1. store the historic events on the same grid as the probabilistic events (res ca 500-1000 Meters), that would also solve the `assign_centroids(...,threshold=0.5)` workaround from the tutorial.
2. integrate `sample_events_from_probs` into the `.set_ls_prob` function and expand it to really create `n_years` events (each representing a year).

Please reach out if you have further questions.
I have already added a few other comments, but I will review again when the changes are implemented

Evelyn-M commented:
Thanks for the feedback so far!
So I made quite some changes to the logic, and hence have not adjusted the unit tests yet.
As discussed I incorporated the following:
- `set_ls_hist` not comes in the same output format as the probabilistic one. It converts the point hazards into raster hazard sets with a user-specifiable resolution. A whole grid cell counts as affected if the point falls within.
- `set_ls_prob` by default produces sampled event-sets now. It is hence of the same logic as the historic one (only that it is properly probabilistic, and not a collection of single events)
- I updated the tutorial accordingly.
- There are 3 helper functions now in the landslide module. 2 for looking up the position of a point on a grid cell matrix, 1 for the binomial sampling from a probability matrix. -  please advice whether they should be transferred to `util`

Thanks :D

chahank commented:
I went through the tutorial and it is very very good. Clear and well written. I have only a few minor suggestions below:

- Remove the
```
import sys
path.climada = ...
...
```
It is not part of other tutorials and should not be necessary in general.

- Set the path to the file according to the default configuration defined in `climada.util.config`. Also, define the paths once as constant (capital letters) instead of defining it several times in the script.

- The import of `entity` is different on the latest develop branch. Probably you should once merge the develop branch into the feature/landslide branch and run the test and tutorial.

- Do not set `reference_year` when setting the exposure from litpop (done twice). Using the default makes it so that the user does not need to download extra night-light data for the arbitrary years 2020 and 2019 used in the tutorial.

- For the note on the historic model and how frequency are set to 1/size_of_set. Maybe add that for `impact.at_event` no frequency correction is made.

- The imports of the landslide module are not needed, so you can remove the lines `import climada.hazard.landslide`

- Could you add the resolution of the probabilistic data to the explanation? 

- I am not sure that the option `check_plots` is good implementation. However, I have to have a look at the code first for that.

I hope these are useful, and stay tuned for the proper code review!

chahank commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
-    def _intensity_prob_to_binom(self, n_years):
-        """convert occurrence probabilities in NGI/UNEP landslide hazard map into binary
-        occurrences (yes/no) within a given time frame.
+    def set_ls_prob(self, bbox, path_sourcefile, corr_fact=10e6, n_years=500, 
```
Maybe the naming of this method should reflect the fact that it is only compatible with the UNEP & NGI dataset.

chahank commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+        fraction stores the actual the occurrence count (0 to n) per grid cell.
+        Frequency is occurrence count / n_years.
+
+        Impact functions, since they act on the intensity, should hence be in
+        the form of a step function,
+        defining impact for intensity 0 and (close to) 1.
```
Why? Could I not want to do an impact function which gives different damages depending on how many occurences there were? (This would require the fraction and intensity to be swapped)

Evelyn-M replied:
mmh no, this should be included anyways, because impact = intensity*MDD*fraction

chahank replied:
Yes, but only as direct addition. This means 2 events always imply twice the damage, which is not necessarily meaningful / desired.

chahank commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+            path to UNEP/NGI ls hazard file (.tif)
+        corr_fact : float or int
+            factor by which to divide the values in the original probability
+            file, in case not scaled to [0,1]. Default is 1'000'000
```
```suggestion
            file, in case it is not scaled to [0,1]. Default is 1'000'000
```

chahank commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
         n_years : int
-            the timespan of the probabilistic simulation in years
+            sampling period
+        check_plots : bool
+            Whether to plot centroids, intensity & fraction of the hazard set.
```
Same as above. Not sure if having plots here is a good idea. The Method already makes several things a once (read data, process data and set data), hence adding also graphs is sub-optimal.

Evelyn-M replied:
fair. i removed it

chahank commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+                                       geometry=[shapely.geometry.box(
+                                           *bbox, ccw=True)])        
+        if "unnamed" in self.centroids.crs.wkt:
+            self.centroids.meta['crs'] = {'init': 'epsg:4326', 'no_defs': True}
```
Is this compatible with the latest version of the develop branch (aka march release?)

Evelyn-M replied:
i changed this syntax. self.centroids.meta['crs'] now evaluates to CRS.from_epsg(4326). is this ok? it is read into this format when using the util.coordinates function read_raster

chahank replied:
hmmm... in `util.coordinates` I find the line `CRS.from_user_input(DEF_CRS)`. Wouldn't this be more suitable (also thinking of future applications with flexible crs choices)?

Evelyn-M replied:
this just explicitly sets the crs in case it is `None` in the file that is read in. Aka it just "blindly" sets the crs to EPSG:4326

chahank replied:
But that is the point. The default use in CLIMADA is `DEF_CRS`. Why not us this default instead of implicitly defining a new one here?

Evelyn-M replied:
changed this now

chahank commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
-                fig2.suptitle('Prob. LS Hazard Set n_years = %i' % n_years, fontsize=14)
-
-            return self
+        return self
```
```suggestion

```

chahank commented on climada/hazard/test/test_landslide.py:
```diff
@@ -19,132 +19,65 @@
+LS_HIST_FILE = DATA_DIR / 'test_ls_hist.shp'
+LS_PROB_FILE = DATA_DIR / 'test_ls_prob.tif'
+
+class TestLandslideModule(unittest.TestCase):     
```
Please add a test for a country that crosses the Anti-meridian (e.g. the Fidji or the US). I am not sure that the bounding boxes in the code can deal with this.

Evelyn-M replied:
True that. it does not.
But in my opinion, this is nothing that has to do with the landslide module, but with the read_raster function, which does the reading of raster files incl bounding boxes of shapely (box) geometries and where this originates from.

I will open an issue on that. We can write sth where, if it a bbox crosses the anti-meridian, it actually reads in two files (left & right of the anti-meridian) and combines them

Evelyn-M replied:
alternatively, i could also just write a wrapper in my landslide module that does that, but i dont find this very clean, since it applies to _any_ raster files that are read in from a bbox

chahank replied:
I agree with your point. It does not make sense to have a special solutions just in this module, but rather the `read_raster` method should be adapted. Opening an issue on this is certainly a good idea.

chahank replied:
However, don't you also have some potential `Landslide` specific issues in the part here 

```
        ls_gdf_bbox = ls_gdf_bbox.assign(row=np.nan, col=np.nan)
        ls_gdf_bbox[['col', 'row']] = ls_gdf_bbox.apply(
            lambda row: mapping_point2grid(row.geometry, bbox[-1], bbox[0], 
                                           res), axis = 1).tolist()
        ls_gdf_bbox['flat_ix'] = ls_gdf_bbox.apply(
            lambda row: mapping_grid2flattened(row.col, row.row, 
                                               self.centroids.shape), axis = 1)        
        self.intensity = sparse.csr_matrix(
            (np.ones(n_ev), (np.arange(n_ev), ls_gdf_bbox.flat_ix)), 
            shape=(n_ev, self.centroids.size))
        self.fraction = self.intensity.copy()
        self.frequency = self.intensity.copy()
```
?

Evelyn-M replied:
don't understand the question

Evelyn-M replied:
took this up in issue #172

chahank commented on climada/hazard/test/test_landslide.py:
```diff
@@ -19,132 +19,65 @@
+        LS_hist = Landslide()
+        LS_hist.set_ls_hist(bbox=(20,40,23,46), 
+                                  path_sourcefile=LS_HIST_FILE, 
+                                  check_plots=0)
```
If the plotting remains, please add a test with the plots.

chahank commented on climada/hazard/test/test_landslide.py:
```diff
@@ -19,132 +19,65 @@
+LS_PROB_FILE = DATA_DIR / 'test_ls_prob.tif'
+
+class TestLandslideModule(unittest.TestCase):     
+
```
Tests for `mapping_point2grid` and `mapping_grid2flattened` are missing.

Evelyn-M replied:
I added them now

chahank commented on climada/hazard/landslide.py:
```diff
@@ -21,160 +21,99 @@
     """
+    return row*matrix_shape[1]+col
+
+def sample_events_from_probs(prob_matrix, n_years):
```
Also, would it not be useful to allow for different kinds of distributions besides the binomial?

Evelyn-M replied:
Yes, I agree. So, would it I) be useful to create a sampling util module? And II), do you think I should already include a kwarg for the type of distribution, in case more options will be implemented under a `sample_events_from_probs` parent function?

chahank replied:
1) Either a sampling util module, or maybe it would also fit in an engine module. Maybe we should a) open an issue on this for an open discussion and b) form a working sub-group at the developer's meeting.
2) I think this could be useful for this particular module, no? Maybe a poisson distribution could be appropriate for sampling the landslide events? In this case, I would say yes, adding a kwarg for the distribution is useful. However, if it is only for a possible future compatibility, I think we can leave it as is because we do not yet really know what the future implementation will look like. Right?

Evelyn-M replied:
as a general remark: I'll open an issue, and for now leave it in the landslide module with two distribution options (binom and poisson)

chahank replied:
I might have missed it, but I do not see the issue on this matter.

chahank reviewed (changes requested):
This is a very clean and clear pull request. Several small comments and a few design questions. Most prominently, should two functions not be made as util functions, should plotting be included in data processing methods, and can the bbox as implemented deal with the anti-meridian?

Cool stuff! Is a publication / applied work planed on this?

chahank commented:
Before I forget, please make sure that the module is compatible with the latest changes to how `crs` are handled. The easiest is probably to merge the develop branch into this branch or rebase your branch on develop (be careful in doing so) and see whether tests and tutorial still run smoothly.

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -21,160 +21,99 @@
-        tiff files (monthly) to current_working_directory/
-            [01-12]_ClimatologyMonthly_032818_9600x5400.tif or
-        tiff files (daily) to save_path/LS_nowcast_date.tif
+def mapping_point2grid(geometry, ymax, xmin, res):
```
Nope, this was exactly my question (see above in my comment to this pull request).

> There are 3 helper functions now in the landslide module. 2 for looking up the position of a point on a grid cell matrix, 1 for the binomial sampling from a probability matrix. - please advice whether they should be transferred to util

Would you advise to put it in the coordinates module?

chahank replied:
Ah yes, I remember, and this confused me. I think this would be a good idea indeed. As a matter of fact, just this morning I came across an application where I needed a function like `mapping_point2grid`.

Evelyn-M replied:
moved it.

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+        ls_gdf_bbox[['col', 'row']] = ls_gdf_bbox.apply(
+            lambda row: mapping_point2grid(row.geometry, bbox[-1],bbox[0], 
+                                           res),
+            axis = 1).tolist()
```
correct!

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
-        for i, j in zip(*self.intensity.nonzero()):
-            self.intensity[i, j] = float((self.intensity[i, j] - min_value) /
-                                         (max_value - min_value) * max_prob)
+        return self
```
so i can just delete the return statement, right?

chahank replied:
Yes. Or return None. Maybe @emanuel-schmid can comment on what is best for CLIMADA consistency?

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+        extent.
+        Points are assigned to the gridcell they fall into, and the whole grid-
+        cell hence counts as equally affected.
+        Event frequencies are roughly estimated by 1/(total time passed in event set),
+        but such estimates are not meaningful and shouldn't be used for 
+        probabilistic calculations! Use the probabilistic method for this!
```
Actually, picking up a comment by you further down, I think it's best to not put any frequency, at all. If one really wanted to have it for some purpose, it's easy to just set this yourself by (no. events / timespan passed in set)

Evelyn-M replied:
I re-changed my mind. climada forces you to put a non-nan frequency. Hence i still think putting a warning and a crappy estimate is better than putting 1 (which implies annual). I'll change it back.

chahank replied:
OK, fair enough.

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+        
+        # assign lat-lon points of LS events to corresponding grid & flattened
+        # grid-index
+        ls_gdf_bbox = ls_gdf_bbox.reindex(columns = ls_gdf_bbox.columns.tolist() + 
```
Adding multiple new empty columns to a dataframe is not super straight-forward, see [here](https://stackoverflow.com/questions/39050539/how-to-add-multiple-columns-to-pandas-dataframe-in-one-assignment)
This was basically solution #7 , i now changed it to #6 with is a bit shorter and still works

chahank replied:
Interesting, I was not aware of this difficulty. In this case two single column additions would also work clean :D. But your call, I just didn't like the reindex which does more than needed.

chahank commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+        
+        self.centroids.set_raster_from_pnt_bounds(bbox,res)
+
+        n_ev = len(ls_gdf_bbox)
```
I see. This is a bit confusing then to have a variable 'bounding box' that is actually an event * locations . This can maybe be made a bit more explicit, but it is not super important.

chahank replied:
I think I still do not understand why a bbox would contain events and not only locations.

Evelyn-M replied:
it's not a bounding box. it's a geodataframe, cut to include only those entries that lie within the bounding box.I could re-name it to `gdf_cropped`or sth if that is more clar.

chahank replied:
yes :D

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -188,304 +127,170 @@ def __init__(self):
+            (minx, miny, maxx, maxy) geographic extent of interest
+         path_sourcefile : str
+             path to shapefile (.shp) with ls point data
+        res : float
+            resolution in degrees of the final grid cells which are created.
+            (1° ~ 111 km at the equator). 
+            Affects the hazard extent! Default is 0.00833
```
espg 4326

Evelyn-M replied:
I dont see how this is related. 1° at the equator is always ca 111km, irrespective of the projection, as long as a full circle is 360°.
And of course, one can manually reproject the gdf if one wanted to, for some reason.

chahank replied:
Ok, then I probably misunderstood something. My concern is the following: with this method, your are dividing the surface in square cells of side lenght defined in ° (the raster). Then, you assign the landslide events to each raster cell. In this case, depending on the projection used, the cells and associated landslide events will vastly differ no?

Evelyn-M replied:
correct. i believe this is the principle of a raster file.

chahank replied:
I don't think so. The idea of a raster file is precisely the opposite. The information included in the raster allows to work with a regular raster on a mathematical level, but there is an implicit projection. Hence, the raster cell size is not independent on the projection. If instead of using degrees, one uses kilometers, the raster file still has the same mathematical regular grid representation, but the square cell side length is completely different.

Evelyn-M replied:
so, i now added in commit 1d0ca7a  the possibility to put in a gdf or a string of where the gdf is located, such that one may re-project the gdf into a more useful projection beforehand. also added some lengthy Note section in the docstring to explain how resolution affects the hazard extent

chahank replied:
Did you add tests for both use cases?

Evelyn-M replied:
yep. that's the test i still wanted to add!

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -21,160 +21,99 @@
-                                 stderr=subprocess.STDOUT, shell=True)
-            p.communicate()
+    for i, j in zip(*prob_matrix.nonzero()):
+        prob_matrix[i, j] = binom.rvs(n=n_years,
```
I changed it now to return a different object, generated initially from a copy of the prob_matrix, on which the trafos are applied

chahank replied:
Ok. If possible in the future, please do not close a comment if changes have been made that should also be reviewed (like in this case).

chahank replied:
Very nice for the reviewer is also if the changes are made in one commit that is referenced in the comments (for larger changes).

Evelyn-M commented on climada/hazard/landslide.py:
```diff
@@ -21,160 +21,99 @@
-            p.communicate()
+    for i, j in zip(*prob_matrix.nonzero()):
+        prob_matrix[i, j] = binom.rvs(n=n_years,
+                                        p=prob_matrix[i, j])
```
yes. changed it now.

Evelyn-M commented:
Alright.
Thanks a lot for the detailed review.
Tests are up and running, so is the tutorial (incl. the convention changes on crs format), and I incorporated the suggestions and / or added my comments on the discussion point. Basically the only things i did _not_ incorporate were including the data source names in the set_ls functions and changing back the order of fraction vs. intensity, since this was the whole point of re-defining intensity and fraction to match the physical background of how the landslides are represented in the data sets.

As such, I'd be ready to merge it. Please let me know if that's ok with you @chahank  @ThomasRoosli .

chahank commented:
There are 2-3 still open points, and I will now have another look at the test files.

chahank commented on climada/hazard/test/test_landslide.py:
```diff
@@ -19,132 +19,65 @@
+        self.assertEqual(max(LS_hist.intensity.data),1)
+
+        
+    def test_set_ls_prob(self):
```
If I am not mistaken, the test with `corr_fact = NotDefaultValue` is missing.

Evelyn-M replied:
added it now.

chahank commented on climada/util/coordinates.py:
```diff
@@ -801,6 +801,64 @@ def get_region_gridpoints(countries=None, regions=None, resolution=150,
+    ------
+    ValueError if Point outside of top left corner of raster
+    """
+    if isinstance(res, tuple):
+        res_x, res_y = (abs(res) for res in res)
+    else:
```
A bit more general, would allow not only tuples but also arrays of np.arrays
```suggestion
    if len(res) == 2:
        res_x, res_y = (abs(res) for res in res)
    else:
```

Evelyn-M replied:
a float object has no length, hence len(res) would not work. that's why i do the type checking. i can add tuple _or_ list in the check

chahank replied:
Aaa yes. Hmmm then the 'pythonic' way would probably be a try...except block ... (better ask for forgiveness than for permission)

chahank commented on climada/util/coordinates.py:
```diff
@@ -801,6 +801,64 @@ def get_region_gridpoints(countries=None, regions=None, resolution=150,
+        res_x, res_y = (abs(res) for res in res)
+    else:
+        res_x = res_y = abs(res)
+    col = int((geometry.x - xmin) / res_x)
+    row = int((ymax - geometry.y) / res_y)
+    if (col < 0 or row < 0):
+        LOGGER.error('Point not inside grid')
+        raise ValueError  
+    return col, row
```
I still have some doubts about this one. 

Is the goal of this function to assign a geographical point (thus, given in a specific coordinate system) to a given grid cell of a raster? Then your function will only work if and only if the raster and the coordinate system are by chance defined in the same coordinate system, which is generally not case right?

Evelyn-M replied:
Sure.
I can put this as a remark in the docstring. 
We have functions in climada to check beforehand whether projections are the same, in case of users needing to check for that.

Evelyn-M replied:
commit 3ce3878 changes geometry to x and y

chahank commented on climada/util/coordinates.py:
```diff
@@ -801,6 +801,64 @@ def get_region_gridpoints(countries=None, regions=None, resolution=150,
+def mapping_grid2flattened(col, row, matrix_shape):
+    """ given a col and row index and the initial 2D matrix shape,
+    return the 1-dimensional index of the same point in the flattened matrix
+    - assumes concatenation in x-direction 
```
```suggestion
    - assumes concatenation along the row-axis (x-direction)
```

chahank reviewed (commented):
Thanks for putting the general function to the util! I think they will be useful to many users/developers. 

I still am not 100% sure whether the `mapping_point2grid` is doing what it should be doing.

Evelyn-M commented:
Alright. Incorporated the two things as discussed yesterday @chahank . Referred to the commits in the comments.

chahank commented:
Basically ready to merge, a few open minor comments that are not critical and one simple test is missing.

Evelyn-M commented:
I added the last test and also incorporated 2 of the minor changes.

chahank commented:
This is ready for merging. I am still not 100% convinced by the choice of how frequency / intensity / fraction are handled. However, the module definitely is clean and working!
